### PR TITLE
feat: support Android protect path

### PR DESCRIPTION
**Pull request by xchacha20-poly1305**
This PR will help to use the portect path, an Android VPN option.

### CODE REVIEW

**Comment by haruue** [comment]
Could you please tell us which application uses such an approach to call `VpnService.protect()`?

**Comment by xchacha20-poly1305** [comment]
> Could you please tell us which application uses such an approach to call `VpnService.protect()`?

Nekobox use it: Code chain: [1](https://github.com/MatsuriDayo/NekoBoxForAndroid/blob/595ed091f933a5359756d08c0ab4ff85f874b8fa/libcore/box.go#L141), [2](https://github.com/MatsuriDayo/libneko/blob/5277a5bfc889ee7a89462695b0e678c1bd4909b1/protect_server/protect_server_linux.go), [3](https://github.com/MatsuriDayo/NekoBoxForAndroid/blob/595ed091f933a5359756d08c0ab4ff85f874b8fa/app/src/main/java/moe/matsuri/nb4a/NativeInterface.kt#L26).
And they used it in [their own fork](https://github.com/MatsuriDayo/hysteria/commit/23223cb5d2759b42646a7da5119c05bdd2543e28)

~~Sing-box has similar logic: code chain [1](https://github.com/SagerNet/sing-box-for-android/blob/2b553d7f7d6587dd095295c7b3ddd74083475d75/app/src/main/java/io/nekohasekai/sfa/bg/VPNService.kt#L47), [2](https://github.com/SagerNet/sing-box/blob/cc8ae725c801419d9926951951816e9ac6ca67d0/experimental/libbox/service.go#L120), [3](https://github.com/SagerNet/sing/blob/3f6c423e76f0c33bcb926b0d69168734951db23f/common/control/protect_unix.go)~~

**Comment by xchacha20-poly1305** [comment]
I tested in my own application, and it is work.

**Comment by haruue** [comment]
Hi there,

I am definitely interested in this idea. However, we need to make some changes to this PR before it can be merged. I've made this PR a draft. When all the work is done on your end, please undraft it to let us know it's ready for review.

File: app/internal/sockopts/sockopts_linux.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+		if err != nil {
+			return fmt.Errorf("failed to create unix socket: %w", err)
+		}
+		defer unix.Close(socketFd)
+
+		timeoutUsec := fdControlUnixTimeout.Microseconds()
+		timeout := unix.Timeval{
+			Sec:  timeoutUsec / 1e6,
+			Usec: timeoutUsec % 1e6,
+		}
```
Comment (xchacha20-poly1305) [inline app/internal/sockopts/sockopts_linux.go:65]: With different architecture, the type of `Sec` and `Usec` are different. Some are `int64` and some are `int32`

**Comment by haruue** [comment]
Both `fwmark` and `bindInterface` are tested and work on my end.

**Review by haruue — APPROVED** [review]

**Comment by haruue** [comment]
@xchacha20-poly1305 

This is still a problem because the socket for the resolver (of the server address) is still un"protect"ed.
 
Android clients might need to resolve the server address and generate a config file with the resolved IP address as `server` and the original hostname (if it is a domain) as `tls.sni`.

**Comment by xchacha20-poly1305** [comment]
~~I felt confused that Hysteria just supported to set DNS for server. And it seemed that we need to refactor many things. Besides, when should we resolve the server address? When creating configuration? When connected? How to set the cache of DNS result?~~

Above this is what I replied before you edited your message.

Fortunately, SagerNet, the first Android GUI supported Hysteria, will automatically add DNS rule for server address (<https://github.com/SagerNet/SagerNet/blob/70e684bae81d4bb4203e860ab88c4319e88f944d/app/src/main/java/io/nekohasekai/sagernet/fmt/ConfigBuilder.kt#L1274-L1282>).

**Review by tobyxdd — APPROVED** [review]

File: app/internal/sockopts/sockopts_linux.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+		if err != nil {
+			return fmt.Errorf("failed to connect: %w", err)
+		}
+
+		err = unix.Sendmsg(socketFd, nil, unix.UnixRights(fd), nil, 0)
+		if err != nil {
+			return fmt.Errorf("failed to send: %w", err)
+		}
+
+		dummy := []byte{1}
```
[the remark this answers is not in the document]
Comment (haruue) [reply app/internal/sockopts/sockopts_linux.go:81]: Furthermore, is there any documentation or protocol specification for the protocol used in this Unix socket? Has it been named? (I plan to refer to it as "fdControl protocol" in our document if not)

**Comment by tobyxdd** [comment]
Hi @xchacha20-poly1305 mind sharing your thoughts on the above questions? We plan to include this in our next release and are currently writing docs for it

**Comment by xchacha20-poly1305** [comment]
@haruue 

Wow, this is an interesting question. I review the history of this protocol, finding it was mentioned by shadowsocks at first. In shadowsocks' Android plugin [documention](https://github.com/shadowsocks/shadowsocks-android/blob/master/plugin/README.md), it's said:

> `__android_vpn`: VPN mode. In this case, the plugin should pass all file descriptors that needs protecting from VPN connections (i.e. its traffic will not be forwarded through the VPN) through an ancillary message to `./protect_path`.

Then I found the first client implement is in [v2ray-plugin](https://github.com/shadowsocks/v2ray-plugin/blob/ddd7ab46b4aeee0ca8b272efed9d7da3e3a6e52c/utils_android.go#L76), calling `ControlOnConnSetup`. And it is not care about `0` or `1`.

And then I checked the implement of ["protect server"](https://github.com/shadowsocks/shadowsocks-android/blob/5a10078d23e930bb6f50de259d4e99ea00459bef/core/src/main/java/com/github/shadowsocks/bg/VpnService.kt#L72), knowning that it will send `0` if failed. So I was confused.

The question of naming it and wheather deal with the responds may ask @madeye for more infomation.

**Pull request merged**